pkg/compact/dedup/reader.go

```diff
@@ -0,0 +1,405 @@
+	lset   labels.Labels
+	chks   []chunks.Meta
+	res    int64
+}
+
+func NewSampleReader(logger log.Logger, cr tsdb.ChunkReader, lset labels.Labels, chks []chunks.Meta, res int64) *SampleReader {
+	return &SampleReader{
+		logger: logger,
+		cr:     cr,
+		lset:   lset,
+		chks:   chks,
+		res:    res,
+	}
+}
+
+func (r *SampleReader) Read(tw *TimeWindow) (map[SampleType][]*Sample, error) {
+	if len(r.chks) == 0 {
+		return nil, nil
+	}
+	if r.res == 0 {
+		return r.readRawSamples(tw)
+	}
+	return r.readDownSamples(tw)
+}
+
+func (r *SampleReader) readRawSamples(tw *TimeWindow) (map[SampleType][]*Sample, error) {
+	samples := make([]*Sample, 0)
+	for _, c := range r.chks {
+		chk, err := r.cr.Chunk(c.Ref)
+		if err != nil {
+			return nil, errors.Wrapf(err, "get raw chunk %d for labels %s", c.Ref, r.lset)
+		}
+		if chk == nil {
+			level.Warn(r.logger).Log("msg", "find empty raw chunk", "ref", c.Ref, "labels", r.lset)
+			continue
+		}
+		ss := r.parseSamples(chk, tw)
+		if len(ss) == 0 {
+			continue
+		}
+		samples = append(samples, ss...)
+	}
+	if len(samples) == 0 {
+		return nil, nil
+	}
+	result := make(map[SampleType][]*Sample)
+	result[RawSample] = samples
+	return result, nil
+}
+
+func (r *SampleReader) readDownSamples(tw *TimeWindow) (map[SampleType][]*Sample, error) {
+	result := make(map[SampleType][]*Sample)
+	result[CountSample] = make([]*Sample, 0)
+	result[SumSample] = make([]*Sample, 0)
+	result[MinSample] = make([]*Sample, 0)
+	result[MaxSample] = make([]*Sample, 0)
+	result[CounterSample] = make([]*Sample, 0)
+	for _, c := range r.chks {
+		chk, err := r.cr.Chunk(c.Ref)
+		if err != nil {
+			return nil, errors.Wrapf(err, "get downsample chunk %d for labels %s", c.Ref, r.lset)
+		}
+		for _, at := range downsampleAggrTypes {
+			ac, err := chk.(*downsample.AggrChunk).Get(at)
+			if err == downsample.ErrAggrNotExist {
+				continue
+			}
+			if ac == nil {
+				level.Warn(r.logger).Log("msg", "find empty downsample chunk", "type", at, "ref", c.Ref, "labels", r.lset)
+				continue
+			}
+			samples := r.parseSamples(ac, tw)
+			if len(samples) == 0 {
+				continue
+			}
+			result[typeMapping[at]] = append(result[typeMapping[at]], samples...)
+		}
+	}
+	return result, nil
+}
+
+func (r *SampleReader) parseSamples(c chunkenc.Chunk, tw *TimeWindow) []*Sample {
+	samples := make([]*Sample, 0)
+	iterator := c.Iterator()
+	for iterator.Next() {
+		timestamp, value := iterator.At()
+		if timestamp < tw.MinTime {
+			continue
+		}
+		// Ignore the data point which timestamp is same with MaxTime.
+		// Make sure the block use scope [MinTime, MaxTime) instead of [MinTime, MaxTime]
+		if timestamp >= tw.MaxTime {
+			break
+		}
+		samples = append(samples, &Sample{
+			timestamp: timestamp,
+			value:     value,
+		})
+	}
+	return samples
+}
+
+type SampleType uint8
+
+const (
+	// samples from non-downsample blocks
+	RawSample SampleType = iota
+	// samples from downsample blocks, map to downsample AggrType
+	CountSample
```

Comment:
Can we just reuse `AggrType`? This will avoid those mappings and simplify code (:

Reply:
Ok, let me figure out the way to handle raw sample with AggrType.